usr/lib/cca_stdll/cca_specific.c

```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
                         (sizeof(cca_mech_list) / sizeof(MECH_LIST_ELEMENT));
 
 
+#ifdef DEBUG
+/* Debug function: hexdump buffer with TRACE_DEBUG */
+static void dump_mem(const char *prestr, void *buf, unsigned len)
+{
+    int i;
+    unsigned char b;
+    char line[70];
+    unsigned char *p = (unsigned char *) buf;
+
+    while (len) {
+        memset(line, ' ', sizeof(line));
+        for (i=0; i < 16; i++) {
+            if (len) {
+                b = *p >> 4;
+                line[3*i+2] = b < 10 ? '0' + b : 'a' - 10 + b;
+                b = *p & 0x0F;
+                line[3*i+3] = b < 10 ? '0' + b : 'a' - 10 + b;
+                line[3*i+4] = ' ';
+                line[51+i] = isalnum(*p) ? *p : '.';
+                p++; len--;
+            }
+        }
+        line[67] = '\0';
+        TRACE_DEBUG("%s%s\n", prestr, line);
+    }
+}
+
+/* Debug function: dump one attribute via TRACE_DEBUG */
+static void dump_attr(CK_ATTRIBUTE *a)
```

Comment:
This should probably go into attributes.c or p11util.c so that it is available for others, too. 
In p11utilc, there are already functions p11_get_ckr() and p11_get_ckm() to get the textual representation of return codes and mechanisms. maybe it would make sense to add a p11_get_cka() as well?